src/PromiseInterface.php

```diff
@@ -82,7 +84,7 @@ public function cancel(): void;
      *
      * If the promise cannot be waited on, then the promise will be rejected.
      *
-     * @return mixed
+     * @return T|PromiseInterface<T>
```

Comment:
I don't think PromiseInterface<T> is entirely correct.
```
/** @var PromiseInterface<int> $promise */
$promise = new FulfilledPromise(2);

/** @var PromiseInterface<bool> $result */
$result = $promise->then(fn ($num) => $num %2 = 0);
```

The return of the callback should determine the new template type.